userland/aero_ipc/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+pub mod ipcmodules {
+    pub use postcard;
+    pub use serde;
+}
+
+/// A MessageHandler is a trait describing an IPC client
+pub trait MessageHandler: Send + Sync {
+    fn handle(&mut self, src: usize, msg: &[u8]) -> Option<Vec<u8>>;
+}
+
+/// A MessageTransport allows for high-level IPC exchanges over the IPC interfce.
+/// It also handles the allocation of request identifiers.
+pub trait MessageTransport {
+    fn alloc_id() -> usize;
+    fn free_id(id: usize);
+    fn exchange(meta: usize, mid: usize, data: &[u8]) -> Vec<u8>;
+}
+
+/// A SendRecieveTransport transfers messages by using the IPC system calls.
+pub struct SendRecieveTransport;
+static IDALLOC: AtomicUsize = AtomicUsize::new(0);
+impl MessageTransport for SendRecieveTransport {
+    fn alloc_id() -> usize {
+        return IDALLOC.fetch_add(1, Ordering::SeqCst);
+    }
+    fn free_id(_: usize) {}
+    fn exchange(meta: usize, mid: usize, msg: &[u8]) -> Vec<u8> {
+        // send the data
+        sys_ipc_send(meta, msg).expect("exchange failed: request failed!");
+        // now wait for a repsonse
+        loop {
+            // get a response
+            let rx = service_with_response_finding();
+            match rx {
+                // if we got a response,
+                Some(mut msg) => {
+                    // and the response has the correct message ID...
+                    let mut deser = postcard::Deserializer::from_bytes(&msg);
+                    let msgid = usize::deserialize(&mut deser)
+                        .expect("message ID not present in the message!");
+                    if msgid == (mid << 1) | 1 {
+                        // return the message contents!
+                        return msg.split_off(8);
+                    }
+                }
+                None => {}
+            }
+        }
+    }
+}
+/// The IPC inteface macro
+///
+/// You can create interfaces like this:
+/// ```no_run
+/// aero_ipc::ipc! {
+///     trait Hello {
+///         fn hello(favorite_number: i32) -> ();
+///     }
+/// ```
+///
+/// Then, Hello::Client is the client interface, Hello::Server is the server
+/// inteface and Hello::handler instantiates a MessageHandler that can be added
+/// to the listening pool.
+#[macro_export]
+macro_rules! ipc {
+    { trait $nm:ident {
+        $(
+            fn $fnnm:ident($($argname:ident : $argty:ty),*) $(-> $t:ty)?;
+        )*
+    } } => {
+        #[allow(non_snake_case)]
+        pub mod $nm {
+            use $crate::ipcmodules::*;
+            pub struct Client<T: $crate::MessageTransport> {
+                pub pid: usize,
+                pub phantom: ::core::marker::PhantomData<T>,
+            }
+            impl<T: $crate::MessageTransport> Client<T> {
+                pub fn pid(&self) -> usize {
+                    self.pid
+                }
+                $(
+                    pub fn $fnnm(&self, $($argname: $argty),*) $(-> $t)? {
```

Comment:
This is up for debate but wouldn't it be better if these functions propagated possible errors instead of crashing the whole program?

Reply:
*crashing the whole program by using `expect` on possible failure types, that will result in a panic which would be very bad in some crucial system service for example

Reply:
oh yeah definitely.

Reply:
at least for the server, crashing it because of a misbehaving process is definitely a denial of service

Reply:
It's true that it's the client misbehaving but we definitely should not let something like that take down possibly a crucial system service, but I'll leave that up for Andy to decide :^)

Reply:
okay, now the server should just print an error but continue running